lane.py

```diff
@@ -618,54 +595,127 @@ def set_list(field_name, value, default=None):
-                best_seller_lane = None
-            if best_seller_lane:
-                self.sublanes.lanes.insert(0, best_seller_lane)
-
+            self.include_best_sellers(**base_args)
 
         # Run some sanity checks.
         ch = Classifier.AUDIENCE_CHILDREN
         ya = Classifier.AUDIENCE_YOUNG_ADULT
         if ((include_best_sellers or include_staff_picks) and
             (self.list_data_source_id or self.list_ids)):
             raise UndefinedLane(
-                "Cannot include best-seller or staff-picks in a lane based on lists."
+                "Cannot include best-seller or staff-picks in a lane "
+                "based on lists."
             )
 
-        if (
-                self.age_range 
-                and not any(x in self.audiences for x in [ch, ya])
-        ):
+        if (self.age_range and
+            not any(x in self.audiences for x in [ch, ya])):
             raise UndefinedLane(
-                "Lane %s specifies age range but does not contain children's or young adult books." % self.name
+                "Lane %s specifies age range but does not contain "
+                "children's or young adult books." % self.name
             )
 
+    def set_from_parent(self, field_name, value, default=None):
+        if value is None:
+            if self.parent:
+                value = getattr(self.parent, field_name, default)
+            else:
+                value = default
+        setattr(self, field_name, value)
+
+    def set_customlist_information(self, list_data_source, list_identifier,
+                                   list_seen_in_previous_days):
+        """Sets any attributes relevant to lanes created from CustomLists"""
+
+        # The lane may be restricted to books that are on a list
+        # from a given data source.
+        custom_list_details = self.custom_lists_for_identifier(
+            list_data_source, list_identifier)
+        (self.list_data_source_id,
+         self.list_ids,
+         self.list_featured_works_query) = custom_list_details
+
+        # Or its parent may be restricted, in which case it should be,
+        # too.
+        if not self.list_data_source_id:
+            self.set_from_parent('list_data_source_id', None)
+        if not self.list_ids:
+            self.set_from_parent('list_ids', None)
+        if not self.list_featured_works_query:
+            # The parent may have featured works from the list that
+            # could be included here.
+            self.set_from_parent('list_featured_works_query', None)
+
+        for sublane in self.sublanes:
+            # If the sublanes were set beforehand, they need to
+            # inherit this information now.
+            sublane.set_customlist_information(None, None, None)
+
+        self.set_from_parent(
+            'list_seen_in_previous_days', list_seen_in_previous_days)
+
+    def custom_lists_for_identifier(self, list_data_source, list_identifier):
```

Comment:
This method and the one below haven't been changed -- just moved up in the file.